staging/src/k8s.io/api/apps/v1/types.go

```diff
@@ -114,6 +114,40 @@ type RollingUpdateStatefulSetStrategy struct {
 	Partition *int32 `json:"partition,omitempty" protobuf:"varint,1,opt,name=partition"`
 }
 
+// PersistentVolumeClaimDeletePolicyType is a string enumeration of the policies that will determine
+// when volumes from the VolumeClaimTemplates will be deleted when the controlling StatefulSet is
+// deleted or scaled down.
+type PersistentVolumeClaimDeletePolicyType string
+
+const (
+	// RetainPersistentVolumeClaimDeletePolicyType is the default
+	// PersistentVolumeClaimDeletePolicy and specifies that
+	// PersistentVolumeClaims associated with StatefulSet VolumeClaimTemplates
+	// will not be deleted.
+	RetainPersistentVolumeClaimDeletePolicyType PersistentVolumeClaimDeletePolicyType = "Retain"
+	// DeletePersistentVolumeClaimDeletePolicyType specifies that
+	// PersistentVolumeClaims associated with StatefulSet VolumeClaimTemplates
+	// will be deleted in the scenario specified in
+	// StatefulSetPersistentVolumeClaimPolicy.
+	DeletePersistentVolumeClaimDeletePolicyType PersistentVolumeClaimDeletePolicyType = "Delete"
+)
+
+// StatefulSetPersistentVolumeClaimPolicy describes the policy used for PVCs
+// created from the StatefulSet VolumeClaimTemplates.
+type StatefulSetPersistentVolumeClaimPolicy struct {
+	// OnSetDeletion specifies what happens to PVCs created from StatefulSet
+	// VolumeClaimTemplates when the StatefulSet is deleted. The default policy
+	// of `Retain` causes PVCs to not be affected by StatefulSet deletion. The
+	// `Delete` policy causes those PVCs to be deleted.
+	OnSetDeletion PersistentVolumeClaimDeletePolicyType `json:"onSetDeletion,omitempty" protobuf:"bytes,1,opt,name=onSetDeletion,casttype=PersistentVolumeClaimDeletePolicyType"`
+	// OnScaleDown specifies what happens to PVCs created from StatefulSet
```

Comment:
OnScaleDown is kind of novel from a naming perspective - are there other field names you drew these from (examples in other workload controllers that use this phrasing)?  I generally would try not to create new name conventions unless it's important - can you point me to either similar existing use or a discussion on the naming?

Reply:
Note the novelty is the primary issue, generally I'd want to ensure the pattern scales first before approving it.

Other options are `WhenX` or `Deleted` (the `On` could be seen as implied).  Just looking for prior art and thought put into it.

Reply:
We chose this in our KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/1847-autoremove-statefulset-pvcs

Is the issue with ScaleDown, or the On* construction? If the former, I guess kubectl scale is what we were inspired by. As for the On* construction, it just read naturally: PersistentVolumeClaimDeletePolicy: { OnSetDeletion: Retain }.

I don't have any other references, though.

Reply:
`On` is the novel bit, although now that you mention it I want to review conventions on what `ScaleDown` means :)

Was there any relevant discussion in the KEP that justified the `On` as opposed to other alternatives?

Reply:
No, not really. We originally had PersistentVolumeClaimDeletePolicy being a field rather than a struct, with possible values DeleteOnStatefulSetDeletion, DeleteOnScaledownOnly, Retain. Then we found a use case for the fourth corner of the matrix, and turned it into a struct, shortening the names to just the On* part.

WhenX makes sense to me as well: WhenScaledDown, WhenSetDeleted. Just "Deleted" seems ambiguous to me for the reasons I mentioned above.

Reply:
we use `when` in the `Pod`. Since this struct belongs to the StatefulSet, I think `whenDeleted` and `whenScaled` is perfectly reasonable (the up/down would be redudant, as only pods are deleted when scaled down). However, maybe the name of `persistentVolumeClaimPolicy` should be `persistentVolumeClaimRetentionPolicy`? Or a double hierarchy if you see other policies coming up in the future.

```golang
type StatefulSetPersistentVolumeClaimPolicy {
  Retention StatefulSetPersistentVolumeClaimRetention
}
type StatefulSetPersistentVolumeClaimRetention {
  WhenDeleted PersistentVolumeClaimRetentionType
  WhenScaled PersistentVolumeClaimRetentionType
}
```

Open for discussion.

Reply:
Re: double hierarchy - I might say for v1 it's ok to skip double level since we don't have a ton of urgent pressure on other policies (unless there's an example I'm missing?).  Technically upgrade is already a policy and we have pod management policy and volume retention is *kind* of like those, so i'd lean towards "no unless we have a really good example now".

Re: `when` good point, i see it in the pod.  I see no examples of `on` in the code base.  Let me see if i can get a quick consensus on the api reviewers mailing list on it and we can enshrine it.

Re: `deleted` will address in other thread.

Reply:
sg, I'll start on when changes as that seems likely outcome and wait for the api reviewers consensus.

Reply:
Waiting just a bit longer for feedback from others, but i think we're leaning towards:

1. `WhenX` (in this context)
2. `On` if it were a hook AND it were a hook that's alone in a mixed struct
3. But most of the time, hooks should be in structs that remove the need for `On` (i.e. in pod lifecycle we don't have OnPreStop).

Reply:
Let's go ahead with these changes (no other feedback came in, I think we're good enough for alpha).

Reply:
Done!